### PR TITLE
Mark record struct generated methods as readonly where possible

### PR DESCRIPTION
**Pull request by AndreyTretyak**
Fixes https://github.com/dotnet/roslyn/issues/54419

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs [hunk 1]
```diff
@@ -30,9 +30,27 @@ internal sealed class SynthesizedRecordDeconstruct : SynthesizedRecordOrdinaryMe
 
         protected override DeclarationModifiers MakeDeclarationModifiers(DeclarationModifiers allowedModifiers, BindingDiagnosticBag diagnostics)
         {
-            const DeclarationModifiers result = DeclarationModifiers.Public;
+            var result = DeclarationModifiers.Public;
+
+            if (ContainingType.IsRecordStruct && !_positionalMembers.Any(m => hasNonReadOnlyGetter(m)))
```
Comment (AndreyTretyak) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs:35]: Looks like `_positionalMembers` is not initialized here for some reason.

Comment (AndreyTretyak) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs]: Had to change the approach to accept `isReadOnly` parameter into the constructor and add a modifier outside of `MakeDeclarationModifiers` method.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs [hunk 2]
```diff
@@ -282,5 +287,43 @@ internal static void VerifyOverridesPrintMembersFromBase(MethodSymbol overriding
                 diagnostics.Add(ErrorCode.ERR_DoesNotOverrideBaseMethod, overriding.Locations[0], overriding, baseType);
             }
         }
+
+        internal static bool AreAllPrintablePropertyGettersReadOnly(NamedTypeSymbol containingType)
+        {
+            return !containingType.GetMembers().Any(m => hasNonReadOnlyGetter(m));
```
Comment (AndreyTretyak) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs:293]: Calling `containingType.GetMembers()` here (executed in base constructor) causes `StackOverflowException` :(

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs]: The `SynthesizedRecordPrintMembers` is constructed when the members of the record are created (in `SourceMemberContainerTypeSymbol`). So we cannot ask about the members of the record in that process...

The way we've sorted this out in the past is that the logic in `SourceMemberContainerTypeSymbol` is phased. We create all the declared members first (those in source) then we create the synthesized ones. The synthesized ones depend on the declared ones, but not vice-versa. This avoids circularity.

I think something like that would work for your purpose: looks at all the declared properties without pulling on all members. That probably needs to be done in `AddSynthesizedRecordMembersIfNecessary` which has access to all declared members and instantiates the `SynthesizedRecordPrintMembers`.

Comment (AndreyTretyak) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs]: Thank you, passing members into constructor from  AddSynthesizedRecordMembersIfNecessary worked. With current implementation we are interested only in user defined properties(since they could have non readonly getters), but it would, probably, be more reliable to check all of them.
I've updated code. I'll check other tests and publish PR if everything is good.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs]: It looks like things are going in the right direction. 

`TopLevelEditingTests.RecordStruct_ImplementSynthesized_PrintMembers` should probably make the explicit implementation use the 'readonly' modifier so that EnC doesn't see it as a rude edit.

`RecordStructTests.Deconstruct_Empty_WithParameterList` should probably update baselines.

`RecordStructTests.ToString_TopLevelRecord_OneField_ConstrainedValueType` shows a bug though.
```

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs]: I took another pass over the spec with the complexity of possibly having to call non-readonly methods on fields in mind. A few observations:

1. for Equals and GetHashCode the spec change is correct as written. We always make a value copy of each field and call into `EqualityComparer<TField>.Default.Equals()/GetHashCode` (which simply calls back into the `TField`'s Equals/GetHashCode if the field's value is not null.). Making the method `readonly` offers a benefit because if the user ever has a readonly ref to the record struct, we won't make a copy of the struct before the call. Therefore fewer copies are made overall.

2. for `PrintMembers()`/`ToString()` it's a little harder to feel out whether the spec change is correct, especially in light of tests like `ToString_TopLevelRecord_OneField_ConstrainedValueType`. However, after some thought I've decided that the spec as written is still the right thing, because it's *already* the behavior for a PrintMembers contained in a readonly record struct. If you inspect the PrintMembers method here, you'll see that the field gets copied the same way: [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4FgBQABAzAAgCcBTCBAexgBsBPI4gY3MIXwGcAXQgVwY/wBKjADwAVAHz4A7gAtiJfKPwh2XXhywBvLPl3489MpVqL8AMyjEqCANxYAvliA=). This is only an issue for the very particular case of fields whose type is a type parameter constrained to struct. For unconstrained type parameters, we simply box and pass to the StringBuilder as an `object`.

To summarize: I think the spec change proposed in dotnet/csharplang#4890 is correct, and the baseline change in `RecordStructTests.ToString_TopLevelRecord_OneField_ConstrainedValueType` is expected.

Comment (AndreyTretyak) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs]: Thank you for the detailed explanations, it's extremely helpful.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 3]
```diff
@@ -3747,7 +3747,12 @@ void addToStringMethod(MethodSymbol printMethod)
                 {
                     if (!memberSignatures.TryGetValue(targetMethod, out Symbol? existingToStringMethod))
                     {
-                        var toStringMethod = new SynthesizedRecordToString(this, printMethod, memberOffset: members.Count, diagnostics);
+                        var toStringMethod = new SynthesizedRecordToString(
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3750]: It occurs to me that it should be specified that the synthesized `ToString` has the same `readonly`-ness as the `PrintMembers`, whether it is synthesized or explicitly declared.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3750]: Have made the change.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs [hunk 4]
```diff
@@ -31,6 +31,7 @@ internal sealed class SynthesizedRecordDeconstruct : SynthesizedRecordOrdinaryMe
         protected override DeclarationModifiers MakeDeclarationModifiers(DeclarationModifiers allowedModifiers, BindingDiagnosticBag diagnostics)
         {
             const DeclarationModifiers result = DeclarationModifiers.Public;
+
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs:34]:
```suggestion
```

nit: remove extra blank line

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 5]
```diff
@@ -3676,7 +3676,7 @@ MethodSymbol addPrintMembersMethod()
                 MethodSymbol printMembersMethod;
                 if (!memberSignatures.TryGetValue(targetMethod, out Symbol? existingPrintMembersMethod))
                 {
-                    printMembersMethod = new SynthesizedRecordPrintMembers(this, memberOffset: members.Count, diagnostics);
+                    printMembersMethod = new SynthesizedRecordPrintMembers(this, generatedMembers: members, userDefinedMembers: userDefinedMembers, diagnostics);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3679]: The reader may be unsettled by a mutable collection being passed to a symbol constructor here. I can see in the implementation that it's not mutated or retained by the symbol, but I think it will probably be best to go back to passing only the `members.Count` here and relying only on the `userDefinedMembers` to determine `isReadOnly`.

To verify our assumption, let's add an assert before `members.AddRange(membersSoFar)` to the effect of `Debug.Assert(isRecordClass || !members.Any(m is PropertySymbol { GetMethod.IsEffectivelyReadOnly: false })`. This would verify that on a record struct, we *never* synthesize a non-`readonly` property.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs [hunk 6]
```diff
@@ -35,6 +35,7 @@ internal abstract class SourceOrdinaryMethodSymbolBase : SourceOrdinaryMethodOrU
             bool isIterator,
             bool isExtensionMethod,
             bool isPartial,
+            bool isReadOnly,
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs:38]: I think this is not the right way to set `isReadOnly` on the synthesized method. Instead of adding a constructor parameter I think we should adjust the value returned by `protected override DeclarationModifiers MakeDeclarationModifiers` in the appropriate synthesized method symbols, e.g. `SynthesizedRecordEquals.MakeDeclarationModifiers`.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs:38]: I'm looking a little closer at the existing `isPartial` and how it is used. It might be reasonable in this design for `SourceOrdinaryMethodSymbol` to check the modifiers in a similar way when passing an argument for it to `SourceOrdinaryMethodSymbolBase`. I will follow up when I'm a little more confident in what the right way is here.

Comment (AndreyTretyak) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs:38]: My initial approach was to change `MakeDeclarationModifiers` to add `readonly` declaration. You can see it here  https://github.com/dotnet/roslyn/pull/55039/files/77be45707a517f0684370ae072a20150f488cfcb
The problem with that is `MakeDeclarationModifiers` being a virtual method still called in the constructor of the base class, so for example in the case with `Deconstruct` `positionalMembers` field are not initialized at the moment of execution.
And as you point out there is a similar problem with `PrintMembers` because we need information about record members.

Would renaming the flag into `IsSynthesizedAsReadOnly` help here? Or should we search for a different approach completely? Check for marking as `readonly` need to be executed before the base constructor, so it does not allow a lot of options. And, I imagine, moving the evaluation of modifiers from the base constructor would be quite problematic as well.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordBaseEquals.cs [hunk 7]
```diff
@@ -17,7 +17,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
     internal sealed class SynthesizedRecordBaseEquals : SynthesizedRecordOrdinaryMethod
     {
         public SynthesizedRecordBaseEquals(SourceMemberContainerTypeSymbol containingType, int memberOffset, BindingDiagnosticBag diagnostics)
-            : base(containingType, WellKnownMemberNames.ObjectEquals, hasBody: true, memberOffset, diagnostics)
+            : base(containingType, WellKnownMemberNames.ObjectEquals, isReadOnly: false, hasBody: true, memberOffset, diagnostics)
```
[the remark this answers is not in the document]
Comment (AndreyTretyak) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordBaseEquals.cs:20]: Makes sense, I've added the check.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordClone.cs [hunk 8]
```diff
@@ -24,7 +24,7 @@ internal sealed class SynthesizedRecordClone : SynthesizedRecordOrdinaryMethod
             SourceMemberContainerTypeSymbol containingType,
             int memberOffset,
             BindingDiagnosticBag diagnostics)
-            : base(containingType, WellKnownMemberNames.CloneMethodName, hasBody: !containingType.IsAbstract, memberOffset, diagnostics)
+            : base(containingType, WellKnownMemberNames.CloneMethodName, isReadOnly: false, hasBody: !containingType.IsAbstract, memberOffset, diagnostics)
```
[the remark this answers is not in the document]
Comment (AndreyTretyak) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordClone.cs:27]: Added in both places, thank you

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 9]
```diff
@@ -3549,6 +3549,9 @@ private void AddSynthesizedRecordMembersIfNecessary(MembersAndInitializersBuilde
             fieldsByName.Free();
             memberNames.Free();
 
+            // Synthesizing non-readonly properties in struct would require changing readonly logic for PrintMembers method synthesis
+            Debug.Assert(isRecordClass || !members.Any(m => m is PropertySymbol { GetMethod.IsEffectivelyReadOnly : false }));
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3553]:
```suggestion
            Debug.Assert(isRecordClass || !members.Any(m => m is PropertySymbol { GetMethod.IsEffectivelyReadOnly: false }));
```
nit: formatting

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs [hunk 10]
```diff
@@ -103,5 +103,22 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
             statementsBuilder.Add(F.Return());
             F.CloseMethod(F.Block(statementsBuilder.ToImmutableAndFree()));
         }
+
+        private static bool IsReadOnly(SourceMemberContainerTypeSymbol containingType, ImmutableArray<Symbol> positionalMembers)
+        {
+            return containingType.IsRecordStruct && !positionalMembers.IsDefault && !positionalMembers.Any(m => hasNonReadOnlyGetter(m));
```
[the remark this answers is not in the document]
Comment (AndreyTretyak) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs]: Good point, I have not thought about it, thanks. Added "fast" path for `PringMembers` and `Deconstruct`.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 5 above)
Comment (AndreyTretyak) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Yes, this would be better. One more reason for that is that the current approach could skip properties generated after `addPrintMembersMethod`. Updated PR, thank you.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 9 above)
Comment (AndreyTretyak) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3553]: For some reason, I'm getting formatting warning about this. Is it a problem with my environment? Or misconfiguration in `.editorconfig`
image

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3553]: Looks like you are experiencing #54759. Let's just go with what the formatting analyzer demands for now.

Comment (AndreyTretyak) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3553]: Looks like it, thanks for finding it. I'm fine with applying the change since it would be correct formatting after analyzer fix (it is also the style for patterns I see in the code), and current formatting with spaces on both sides looks strange :)

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: The trouble with applying the change is that some of the CI legs will fail due to presence of warnings :/

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs (hunk 6 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs:38]: Thanks for the info. You're correct that my initial suggestion doesn't result in the correct semantics.

I think for now what I'd like to do is add an additional assertion to `SourceOrdinaryMethodSymbolBase` to make it clear that this isn't how explicitly declared methods are made `readonly`:

```cs
Debug.Assert(!isReadOnly || this.IsImplicitlyDeclared,
    "We only expect synthesized methods to use this flag to make a method readonly. Explicitly declared methods should get this value from modifiers in syntax.");
```

I think in the future I'd like to make an attempt to push the virtual call out of the constructor, perhaps into the constructor arguments. But this PR doesn't seem like the right time to make such an attempt.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 9 above)
Comment (AndreyTretyak) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Oh, sorry, I'll revert it back.

**Comment by RikkiGibson** [comment]
I also wanted to note that I benchmarked ToString performance when the method was readonly vs non-readonly. The first table is from a simpler "non-generic scenario. The second scenario tries to account for the impact of copying fields constrained to struct within the `ToString` before calling the field's `ToString`.

### Non-generic
``` ini
BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19042.1110 (20H2/October2020Update)
AMD Ryzen 9 3900X, 1 CPU, 24 logical and 12 physical cores
.NET SDK=6.0.100-preview.6.21355.2
  [Host]     : .NET 6.0.0 (6.0.21.35212), X64 RyuJIT
  DefaultJob : .NET 6.0.0 (6.0.21.35212), X64 RyuJIT


|                       Method |     Mean |   Error |  StdDev |
|----------------------------- |---------:|--------:|--------:|
|                     ToString | 338.5 ns | 1.14 ns | 1.01 ns |
|             ToStringReadonly | 273.8 ns | 1.63 ns | 1.53 ns |
|         ToString_RefReadonly | 345.6 ns | 4.22 ns | 3.74 ns |
| ToStringReadonly_RefReadonly | 265.8 ns | 1.11 ns | 0.98 ns |
```

### Generic `where T : struct`
``` ini
BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19042.1110 (20H2/October2020Update)
AMD Ryzen 9 3900X, 1 CPU, 24 logical and 12 physical cores
.NET SDK=6.0.100-preview.6.21355.2
  [Host]     : .NET 6.0.0 (6.0.21.35212), X64 RyuJIT
  DefaultJob : .NET 6.0.0 (6.0.21.35212), X64 RyuJIT


|                       Method |     Mean |   Error |  StdDev |
|----------------------------- |---------:|--------:|--------:|
|                     ToString | 332.1 ns | 1.66 ns | 1.39 ns |
|             ToStringReadonly | 283.6 ns | 2.63 ns | 2.46 ns |
|         ToString_RefReadonly | 338.8 ns | 1.45 ns | 1.21 ns |
| ToStringReadonly_RefReadonly | 267.2 ns | 0.76 ns | 0.64 ns |
```

<details>
    <summary>Benchmark source</summary>

```cs
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System;
using System.Runtime.CompilerServices;
using System.Text;

namespace bench_readonly
{
    class Program
    {
        static void Main(string[] args)
        {
            var summary = BenchmarkRunner.Run(typeof(Program).Assembly);
        }
    }

    [MarkdownExporterAttribute.GitHub]
    public class ToStringFix
    {
        [Benchmark]
        public new void ToString()
        {
            var rec1 = new Rec1("abc", false, 42, new Rec2("def", 123));
            rec1.ToString();
        }

        [Benchmark]
        public void ToStringReadonly()
        {
            var rec3 = new Rec3("abc", false, 42, new Rec4("def", 123));
            rec3.ToString();
        }

        [Benchmark]
        public void ToString_RefReadonly()
        {
            var rec1 = new Rec1("abc", false, 42, new Rec2("def", 123));
            local(in rec1);

            void local(in Rec1 rec1) => rec1.ToString();
        }

        [Benchmark]
        public void ToStringReadonly_RefReadonly()
        {
            var rec3 = new Rec3("abc", false, 42, new Rec4("def", 123));
            local(in rec3);

            void local(in Rec3 rec3) => rec3.ToString();
        }

        public record struct Rec1(string Prop1, bool Prop2, int Prop3, Rec2 Prop4);
        public record struct Rec2(string Prop1, int Prop2);

        public record struct Rec3(string Prop1, bool Prop2, int Prop3, Rec4 Prop4)
        {
            private readonly bool PrintMembers(StringBuilder builder)
            {
                builder.Append("Prop1 = ");
                builder.Append(Prop1);
                builder.Append(", Prop2 = ");
                builder.Append(Prop2.ToString());
                builder.Append(", Prop3 = ");
                builder.Append(Prop3.ToString());
                builder.Append(", Prop4 = ");
                builder.Append(Prop4);
                return true;
            }

            public readonly override string ToString()
            {
                var builder = new StringBuilder();
                builder.Append("Rec3 { ");
                if (PrintMembers(builder)) builder.Append(' ');
                builder.Append('}');
                return builder.ToString();
            }
        }
        public record struct Rec4(string Prop1, int Prop2)
        {
            private readonly bool PrintMembers(StringBuilder builder)
            {
                builder.Append("Prop1 = ");
                builder.Append(Prop1);
                builder.Append(", Prop2 = ");
                builder.Append(Prop2.ToString());
                return true;
            }

            public readonly override string ToString()
            {
                var builder = new StringBuilder();
                builder.Append("Rec4 { ");
                if (PrintMembers(builder)) builder.Append(' ');
                builder.Append('}');
                return builder.ToString();
            }
        }
    }

    [MarkdownExporterAttribute.GitHub]
    public class ToStringFix_StructConstrainedField
    {
        [Benchmark]
        public new void ToString()
        {
            var rec1 = new Rec1<Rec2>("abc", false, 42, new Rec2("def", 123));
            rec1.ToString();
        }

        [Benchmark]
        public void ToStringReadonly()
        {
            var rec3 = new Rec3<Rec4>("abc", false, 42, new Rec4("def", 123));
            rec3.ToString();
        }

        [Benchmark]
        public void ToString_RefReadonly()
        {
            var rec1 = new Rec1<Rec2>("abc", false, 42, new Rec2("def", 123));
            local(in rec1);

            void local(in Rec1<Rec2> rec1) => rec1.ToString();
        }

        [Benchmark]
        public void ToStringReadonly_RefReadonly()
        {
            var rec3 = new Rec3<Rec4>("abc", false, 42, new Rec4("def", 123));
            local(in rec3);

            void local(in Rec3<Rec4> rec3) => rec3.ToString();
        }

        public record struct Rec1<T>(string Prop1, bool Prop2, int Prop3, T Prop4) where T : struct;
        public record struct Rec2(string Prop1, int Prop2);

        public record struct Rec3<T>(string Prop1, bool Prop2, int Prop3, T Prop4) where T : struct
        {
            private readonly bool PrintMembers(StringBuilder builder)
            {
                builder.Append("Prop1 = ");
                builder.Append(Prop1);
                builder.Append(", Prop2 = ");
                builder.Append(Prop2.ToString());
                builder.Append(", Prop3 = ");
                builder.Append(Prop3.ToString());
                builder.Append(", Prop4 = ");
                builder.Append(Prop4);
                return true;
            }

            public readonly override string ToString()
            {
                var builder = new StringBuilder();
                builder.Append("Rec3 { ");
                if (PrintMembers(builder)) builder.Append(' ');
                builder.Append('}');
                return builder.ToString();
            }
        }
        public record struct Rec4(string Prop1, int Prop2)
        {
            private readonly bool PrintMembers(StringBuilder builder)
            {
                builder.Append("Prop1 = ");
                builder.Append(Prop1);
                builder.Append(", Prop2 = ");
                builder.Append(Prop2.ToString());
                return true;
            }

            public readonly override string ToString()
            {
                var builder = new StringBuilder();
                builder.Append("Rec4 { ");
                if (PrintMembers(builder)) builder.Append(' ');
                builder.Append('}');
                return builder.ToString();
            }
        }
    }
}
```
</details>

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 9 above)
Comment (AndreyTretyak) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3553]: Reverted

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs (hunk 6 above)
Comment (AndreyTretyak) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs:38]: Thanks, I've added assert. Do you mind if I try to move the virtual call out after this PR?

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs (hunk 6 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs:38]: Feel free to give it a try.

**Comment by AndreyTretyak** [comment]
@RikkiGibson  It looks to me that pipeline failure is unrelated to this change. Is it possible and if so what would be recommended actions? Wait for a fix in `main` and merge changes into PR?

**Comment by RikkiGibson** [comment]
Let's just try re-running once more, if they still fail, then let's look at merging 'main' into the PR branch.

**Comment by RikkiGibson** [comment]
@jcouv @dotnet/roslyn-compiler for second review.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs [hunk 11]
```diff
@@ -44,6 +45,7 @@ internal abstract class SourceOrdinaryMethodSymbolBase : SourceOrdinaryMethodOrU
                  isIterator: isIterator)
         {
             Debug.Assert(diagnostics.DiagnosticBag is object);
+            Debug.Assert(!isReadOnly || this.IsImplicitlyDeclared, "We only expect synthesized methods to use this flag to make a method readonly. Explicitly declared methods should get this value from modifiers in syntax.");
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs:48]: Consider also asserting that we ever only set `isReadOnly` inside a struct `containingType`.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 26)

**Comment by RikkiGibson** [comment]
Thanks for the contribution @AndreyTretyak!
